### PR TITLE
Move CMAC to EVP_MAC

### PR DESCRIPTION
**Pull request by levitte**
Add CMAC implementation
Integrate CMAC EVP_PKEY_METHOD into generic MAC EVP_PKEY_METHOD
Adapt other EVP code to use EVP_MAC instead of direct implementation calls

### CODE REVIEW

**Comment by levitte** [comment]
Updated and documentation added.

**Comment by levitte** [comment]
Now that #7393 is in, I rebased this on top of master.

File: crypto/cmac/cm_meth.c [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+}
+
+static int cmac_init(EVP_MAC_IMPL *cctx)
+{
+    int rv = CMAC_Init(cctx->ctx, NULL, 0, cctx->tmpcipher,
+                       (ENGINE *)cctx->tmpengine);
+    cctx->tmpcipher = NULL;
+    cctx->tmpengine = NULL;
+
+    if (!rv)
```
Comment (paulidale) [inline crypto/cmac/cm_meth.c]: Would `return rv;` work here?  `CMAC_Init` returns 0 or 1.
Or even `return rv ? 1 : 0;` ?

Similarly for the next couple of functions.

Comment (levitte) [reply crypto/cmac/cm_meth.c]: Sure.

Comment (paulidale) [reply crypto/cmac/cm_meth.c]: The other macs (HMAC, ...) had similar.

**Review by paulidale — APPROVED** [review]
The plumbing for MACs looks like it much easier now.

**Comment by levitte** [comment]
> The plumbing for MACs looks like it much easier now.

That was the intention.

**Comment by levitte** [comment]
Thanks for the review!

Merged.

56adb7d93721a72bfae532845cbebc4a565ceb65 Make sure at least one CMAC test still uses the EVP_PKEY method
b8d77c9bd675b4128aeeafb4a738938460477a2e Adapt other EVP code to use EVP_MAC instead of direct implementation calls
e74a435f58441c6f1f6b4558c762e17d0ab67b7f EVP_MAC: Integrate CMAC EVP_PKEY_METHOD into generic MAC EVP_PKEY_METHOD
f71faf2753cc1b1cbba0da0997b70e5a908ac24b EVP_MAC: Add CMAC implementation

**Pull request closed without merging**